Erased constructors (#4638)

Saizan opened the pull request:
PR for #4638

Saizan committed:
[ re #4638 ] Allow erased constructors

and check modality for constructors application too.

Saizan committed:
[ re #4638 ] LHS: Propagate quantity to result type

Saizan committed:
[ re #4638 ] Coverage: erased env if splitting on an erased constructor

andreasabel commented on src/full/Agda/TypeChecking/Coverage.hs:
```diff
@@ -930,7 +930,20 @@ fixTargetType sc@SClause{ scTel = sctel, scSubst = sigma } target = do
+        SplitCon c -> do
+          q <- getQuantity <$> getConstInfo c
+          case q of
+            Quantity0{} -> return (q <>)
```
Picky comment: Shouldn't this rather return `mapQuantity (q <>)` and then below use just `f` instead of `mapQuantity f`?

andreasabel reviewed (changes requested):
Do we need to erase erased clauses during compilation or is this happening already?

Saizan committed:
[ re #4638 ] tweak

Saizan commented:
> Do we need to erase erased clauses during compilation or is this happening already?

That's a good question, is there a good way to setup a test for that?

Meanwhile I'll try to look at the compiler code.

Saizan committed:
[ re #4638 ] single non-erased constructor @0 matching

Saizan committed:
[ re #4638 ] actually erase constructors in the MAlonzo backend.

Saizan committed:
[ re #4638 ] fix build

nad commented:
A non-trivial test case that currently fails:
```agda
{-# OPTIONS --cubical #-}

module Erased-constructor-test where

open import Agda.Builtin.Char
open import Agda.Builtin.String
open import Equality.Path
import Equality.Propositional as EP
open import IO-monad
open import Prelude
open import String

open import Bag-equivalence equality-with-J as BE using (_∼[_]_; bag)
open import List equality-with-J as L
open import Monad EP.equality-with-J
open import H-level.Closure equality-with-J
open import Quotient.Erased equality-with-paths as Q

len : List Char /ᴱ _∼[ bag ]_ → ℕ
len = Q.rec λ where
  .[]ʳ                   → length
  .[]-respects-relationʳ → BE.length-cong
  .is-setʳ               → ℕ-set

total-len : List String → ℕ
total-len = sum ∘ L.map (len ∘ [_] ∘ primStringToList)

main = do
  args ← getArgs
  putStrLn (prettyℕ (total-len args))
```
To test:
```
git clone --single-branch --branch erased-constructors https://github.com/nad/equality.git
```
Save `Erased-constructor-test.agda` in `equality`.
```
cd equality
agda -i. --compile Erased-constructor-test.agda 
```
Currently Agda complains:
```
An internal error has occurred. Please report this as a bug.
Location of the error: src/full/Agda/Compiler/MAlonzo/Primitives.hs:138
```
I guess this is expected, because I have not marked all cubical code as erased. However, I think that all cubical code reachable from `main` is marked as erased.

Saizan commented:
You should just get a `NotImplemented` or runtime error if some cubical primitive ends up in the compiled code.

The impossible is Malonzo not even being able to get at the builtin name. I'll test on my end.

nad commented:
Before this is released we should try to make sure that this kind of code does not lead to run-time errors.

Saizan committed:
[ re #4638 ] fix getPrimName and remove the NotImplemented error for now.

Saizan committed:
[ compiler ] do not inline extended lambdas with recursive clauses

It caused the compiler to loop.

Saizan committed:
[ compiler cubical ] do not crash on hcomp splits and path constructors

nad commented:
I fixed some problems related to #4727 by manually giving some implicit arguments (and force-pushed the changes). Now I get the following error:
```
MAlonzo/Code/HZ45Zlevel/Truncation/Propositional.hs:214:7: error:
    • The constructor ‘C1720’ should have 2 arguments, but has been given 3
```
The eliminator for the propositional truncation operator with a *non-erased* higher constructor is not compiled correctly. Perhaps it makes sense to (optionally) only compile code that can be reached from `main`.

I'll try to do some manual pruning.

nad commented:
```
$ ./Erased-constructor-test It works
7
```

nad commented:
I've pushed a commit with the manually pruned code to the branch, so that you can also test this.

Saizan commented:
@nad did you say you had merged master into this?

nad commented:
I've pushed some merge commits now.

nad committed:
[ #4638 ] Fixed a test case.

nad committed:
[ #4784 ] Use correct (?) quantities when checking constructors.

Developed in collaboration with Andrea Vezzosi.

nad commented:
One compiler test (#2918) fails.

nad committed:
[ #4638 ] The GHC and JS backends are now more picky.

Both refuse to compile higher inductive types, as well as (at least
some) cubical primitives.

nad committed:
Merged master into issue4638.

nad committed:
[ #2918, #4638 ] Disabled a test case.

nad committed:
[ #4638 ] Tests.

nad commented:
> Before this is released we should try to make sure that this kind of code does not lead to run-time errors.

I've made the backends more picky. Now they refuse to compile higher inductive types and (at least some) cubical primitives.
